# DocFrame Demo: Text Analysis Made Simple

Welcome to DocFrame! This notebook demonstrates how DocFrame makes text analysis in Python as simple and powerful as possible by combining the performance of Polars with intelligent text processing capabilities.

## What is DocFrame?

DocFrame extends Polars DataFrames with document-aware functionality, automatically detecting text columns and providing unified text processing APIs. It bridges the gap between raw data manipulation and text analysis.

### Key Features:
- 🚀 **Automatic Document Detection**: Intelligently identifies your main text column
- 🔄 **Unified Text API**: Same text operations work on Series, DataFrames, and Expressions  
- ⚡ **Performance**: Built on Polars for speed and memory efficiency
- 🎯 **Seamless Integration**: Drop-in replacement that enhances your existing Polars workflows
- 📊 **Rich Text Analytics**: Built-in text statistics, cleaning, and transformation tools

Let's explore using real political data from the 2020 Queensland election!

## 1. Setup and Imports

In [1]:
import polars as pl
import docframe
candidates_df = pl.read_csv("data/ADO/candidate_info_gender.csv")
tweets_df = docframe.read_csv("data/ADO/qldelection2020_candidate_tweets.csv")
tweets_df

DocDataFrame(shape: (2_380, 15)
┌───────────┬───────────┬───────────┬───────────┬───┬───────────┬───────────┬───────────┬──────────┐
│ tweet_id  ┆ created_a ┆ retrieved ┆ user_id   ┆ … ┆ in_reply_ ┆ quoted_tw ┆ quoted_us ┆ quoted_u │
│ ---       ┆ t         ┆ _at       ┆ ---       ┆   ┆ to_user_n ┆ eet_id    ┆ er_id     ┆ ser_name │
│ i64       ┆ ---       ┆ ---       ┆ i64       ┆   ┆ ame       ┆ ---       ┆ ---       ┆ ---      │
│           ┆ str       ┆ str       ┆           ┆   ┆ ---       ┆ i64       ┆ i64       ┆ str      │
│           ┆           ┆           ┆           ┆   ┆ str       ┆           ┆           ┆          │
╞═══════════╪═══════════╪═══════════╪═══════════╪═══╪═══════════╪═══════════╪═══════════╪══════════╡
│ 131712322 ┆ 2020-10-1 ┆ 2020-10-1 ┆ 162876955 ┆ … ┆ null      ┆ null      ┆ null      ┆ null     │
│ 653577625 ┆ 6 15:20:2 ┆ 7 02:01:4 ┆           ┆   ┆           ┆           ┆           ┆          │
│ 7         ┆ 2.000     ┆ 6.000     ┆           ┆   ┆           ┆           ┆           ┆          │
│           ┆ +0000     ┆ +0000     ┆           ┆   ┆           ┆           ┆           ┆          │
│ 131712348 ┆ 2020-10-1 ┆ 2020-10-1 ┆ 162876955 ┆ … ┆ null      ┆ null      ┆ null      ┆ null     │
│ 283708620 ┆ 6 15:21:2 ┆ 7 02:01:4 ┆           ┆   ┆           ┆           ┆           ┆          │
│ 8         ┆ 3.000     ┆ 6.000     ┆           ┆   ┆           ┆           ┆           ┆          │
│           ┆ +0000     ┆ +0000     ┆           ┆   ┆           ┆           ┆           ┆          │
│ 131712356 ┆ 2020-10-1 ┆ 2020-10-1 ┆ 162876955 ┆ … ┆ null      ┆ null      ┆ null      ┆ null     │
│ 424855142 ┆ 6 15:21:4 ┆ 7 02:01:4 ┆           ┆   ┆           ┆           ┆           ┆          │
│ 5         ┆ 3.000     ┆ 6.000     ┆           ┆   ┆           ┆           ┆           ┆          │
│           ┆ +0000     ┆ +0000     ┆           ┆   ┆           ┆           ┆           ┆          │
│ 131712370 ┆ 2020-10-1 ┆ 2020-10-1 ┆ 162876955 ┆ … ┆ null      ┆ null      ┆ null      ┆ null     │
│ 467801907 ┆ 6 15:22:1 ┆ 7 02:01:4 ┆           ┆   ┆           ┆           ┆           ┆          │
│ 2         ┆ 6.000     ┆ 6.000     ┆           ┆   ┆           ┆           ┆           ┆          │
│           ┆ +0000     ┆ +0000     ┆           ┆   ┆           ┆           ┆           ┆          │
│ 131712415 ┆ 2020-10-1 ┆ 2020-10-1 ┆ 162876955 ┆ … ┆ null      ┆ null      ┆ null      ┆ null     │
│ 024315187 ┆ 6 15:24:0 ┆ 7 02:01:4 ┆           ┆   ┆           ┆           ┆           ┆          │
│ 4         ┆ 3.000     ┆ 6.000     ┆           ┆   ┆           ┆           ┆           ┆          │
│           ┆ +0000     ┆ +0000     ┆           ┆   ┆           ┆           ┆           ┆          │
│ …         ┆ …         ┆ …         ┆ …         ┆ … ┆ …         ┆ …         ┆ …         ┆ …        │
│ 132336374 ┆ 2020-11-0 ┆ 2020-11-0 ┆ 129802370 ┆ … ┆ Annastaci ┆ 132318155 ┆ 256833444 ┆ Annastac │
│ 386293555 ┆ 2 20:37:5 ┆ 2 21:45:2 ┆ 885905203 ┆   ┆ aMP       ┆ 230251417 ┆           ┆ iaMP     │
│ 2         ┆ 8.000     ┆ 1.000     ┆ 2         ┆   ┆           ┆ 6         ┆           ┆          │
│           ┆ +0000     ┆ +0000     ┆           ┆   ┆           ┆           ┆           ┆          │
│ 132336410 ┆ 2020-11-0 ┆ 2020-11-0 ┆ 452670762 ┆ … ┆ null      ┆ null      ┆ null      ┆ null     │
│ 633472819 ┆ 2 20:39:2 ┆ 2 21:15:1 ┆           ┆   ┆           ┆           ┆           ┆          │
│ 3         ┆ 4.000     ┆ 9.000     ┆           ┆   ┆           ┆           ┆           ┆          │
│           ┆ +0000     ┆ +0000     ┆           ┆   ┆           ┆           ┆           ┆          │
│ 132336613 ┆ 2020-11-0 ┆ 2020-11-0 ┆ 452670762 ┆ … ┆ null      ┆ null      ┆ null      ┆ null     │
│ 359699968 ┆ 2 20:47:2 ┆ 2 21:15:1 ┆           ┆   ┆           ┆           ┆           ┆          │
│ 0         ┆ 7.000     ┆ 9.000     ┆           ┆   ┆           ┆           ┆           ┆          │
│           ┆ +0000     ┆ +00

In [3]:
tweets_df.group_by('username').agg([
    pl.col('text').count(),]).sort('text')

shape: (79, 2)
┌─────────────────┬──────┐
│ username        ┆ text │
│ ---             ┆ ---  │
│ str             ┆ u32  │
╞═════════════════╪══════╡
│ meaghan_scanlon ┆ 1    │
│ Ros_Bates_MP    ┆ 1    │
│ Bennett4Burnett ┆ 1    │
│ CraigCrawfordMP ┆ 1    │
│ JGilbertMP      ┆ 1    │
│ …               ┆ …    │
│ camerondickqld  ┆ 107  │
│ mireelr         ┆ 119  │
│ BrunkerMichael  ┆ 172  │
│ MarkBaileyMP    ┆ 311  │
│ AnnastaciaMP    ┆ 315  │
└─────────────────┴──────┘

In [5]:
tweets_df_AnnastaciaMP = tweets_df.filter(pl.col("username") == "AnnastaciaMP")
tweets_df_MarkBaileyMP = tweets_df.filter(pl.col("username") == "MarkBaileyMP")
docframe.compute_token_frequencies({'tweets_df_AnnastaciaMP': tweets_df_AnnastaciaMP, 'tweets_df_MarkBaileyMP': tweets_df_MarkBaileyMP})

{'tweets_df_AnnastaciaMP': {'0': 12,
  '1': 16,
  '10': 3,
  '100': 1,
  '1000': 2,
  '100k': 1,
  '100m': 1,
  '1015': 0,
  '10km': 0,
  '10m': 0,
  '10newsfirstqld': 0,
  '10year': 2,
  '11': 1,
  '1100': 0,
  '1139': 2,
  '1154': 5,
  '1155': 3,
  '1156': 1,
  '1157': 3,
  '1159': 2,
  '1160': 2,
  '1163': 1,
  '1164': 4,
  '1165': 2,
  '1167': 5,
  '1169': 1,
  '1171': 2,
  '1172': 3,
  '1184163': 1,
  '1186578': 1,
  '1188355': 1,
  '1192435': 1,
  '1198424': 1,
  '12': 1,
  '120': 1,
  '1203184': 1,
  '1207967': 1,
  '120m': 0,
  '1211609': 1,
  '1214121': 1,
  '1215772': 1,
  '1219570': 1,
  '1224099': 1,
  '1229045': 1,
  '1233236': 1,
  '1236847': 1,
  '1239418': 1,
  '124': 0,
  '1241130': 1,
  '126': 1,
  '126b': 0,
  '13': 1,
  '137': 0,
  '13b': 1,
  '14': 1,
  '140': 4,
  '14000': 0,
  '15': 2,
  '150000': 0,
  '154': 0,
  '15b': 0,
  '16': 1,
  '160': 1,
  '16b': 0,
  '16km': 0,
  '17': 1,
  '171': 1,
  '17800': 1,
  '19': 2,
  '193': 1,
  '193m': 1,
  '1am': 1,
  '1b': 

In [19]:
# Essential imports
import polars as pl
import docframe  # This registers the text namespace
from docframe import DocDataFrame, DocLazyFrame

# Optional: for better display formatting
import warnings

warnings.filterwarnings("ignore")

print("DocFrame is ready! 🚀")
print(f"Polars version: {pl.__version__}")
print("Text namespace registered for enhanced text processing")

# Quick validation that text namespace is available
sample_df = pl.DataFrame({"text": ["hello world", "test text"]})
print(f"Text namespace available: {hasattr(sample_df.get_column('text'), 'text')}")

DocFrame is ready! 🚀
Polars version: 1.31.0
Text namespace registered for enhanced text processing
Text namespace available: True


## 2. Loading Real-World Data

We'll use two datasets from the 2020 Queensland election:
- **Candidate Information**: Details about political candidates including gender
- **Tweet Data**: Over 6,000 tweets from candidates during the campaign

Let's see how DocFrame automatically detects the main text columns!

In [20]:
# Load candidate information data
candidates_df = docframe.read_csv("data/ADO/candidate_info_gender.csv")

print("🗳️ Candidate Information Dataset")
print(f"Shape: {candidates_df.shape}")
print(f"Automatically detected document column: '{candidates_df.active_document_name}'")
print(f"Available columns: {candidates_df.columns}")
print()

# Show first few rows
print("Sample data:")
print(
    candidates_df.head(3).select(
        ["party", "first_name", "last_name", "gender", "evidence"]
    )
)
print()

# Let's see what the evidence column (auto-detected as document) contains
print("Evidence column samples (auto-detected document column):")
for i, text in enumerate(candidates_df.document.head(3).to_list()):
    print(f"{i + 1}. {text[:100]}...")

🗳️ Candidate Information Dataset
Shape: (133, 8)
Automatically detected document column: 'evidence'
Available columns: ['party', 'electorate', 'first_name', 'last_name', 'twitter_url', 'username', 'gender', 'evidence']

Sample data:
Document column: 'evidence'
shape: (3, 5)
┌──────────────────────────┬────────────┬───────────┬────────┬─────────────────────────────────┐
│ party                    ┆ first_name ┆ last_name ┆ gender ┆ evidence                        │
│ ---                      ┆ ---        ┆ ---       ┆ ---    ┆ ---                             │
│ str                      ┆ str        ┆ str       ┆ str    ┆ str                             │
╞══════════════════════════╪════════════╪═══════════╪════════╪═════════════════════════════════╡
│ Informed Medical Options ┆ Marilyn    ┆ Winters   ┆ F      ┆ "THREE Redlands women from the… │
│ (IMOP…                   ┆            ┆           ┆        ┆                                 │
│ Informed Medical Options ┆ Tara       ┆ Garo

In [18]:
# Load tweet data - this is the main dataset for text analysis
tweets_df = docframe.read_csv("data/ADO/qldelection2020_candidate_tweets.csv")

print("🐦 Tweet Dataset")
print(f"Shape: {tweets_df.shape}")
print(f"Automatically detected document column: '{tweets_df.active_document_name}'")
print(f"Available columns: {tweets_df.columns}")
print()

# Show sample tweets
print("Sample tweets:")
for i, tweet in enumerate(tweets_df.document.head(3).to_list()):
    print(f"{i + 1}. {tweet}")
    print("   " + "-" * 80)
print()

# Show dataset info using DocFrame built-in method
print("Dataset Overview:")
print(f"Total tweets: {len(tweets_df):,}")
print(f"Unique users: {tweets_df.select(pl.col('username')).n_unique()}")
print(
    f"Date range: {tweets_df.select(pl.col('created_at').min(), pl.col('created_at').max())}"
)

🐦 Tweet Dataset
Shape: (2380, 15)
Automatically detected document column: 'text'
Available columns: ['tweet_id', 'created_at', 'retrieved_at', 'user_id', 'username', 'text', 'retweeted_tweet_id', 'retweeted_user_id', 'retweeted_user_name', 'in_reply_to_tweet_id', 'in_reply_to_user_id', 'in_reply_to_user_name', 'quoted_tweet_id', 'quoted_user_id', 'quoted_user_name']

Sample tweets:
1. RT @GhostWhoVotes: #Newspoll QLD State 2 Party Preferred: ALP 52 (+3 since July) LNP 48 (-3) #qldvotes #auspol
   --------------------------------------------------------------------------------
2. RT @camerondickqld: And @DebFrecklington &amp; @TimManderMP  your ideologically-driven obsession with a surplus will lead to significant economic problems - 30,000 Queenslanders sacked, sky-high unemployment, health services cut, regional communities punished &amp; a longer and deeper recession.
   --------------------------------------------------------------------------------
3. RT @QLDLabor: Different electi

DuplicateError: the name 'created_at' is duplicate

It's possible that multiple expressions are returning the same default column name. If this is the case, try renaming the columns with `.alias("new_name")` to avoid duplicate column names.

## 3. DocFrame Text Namespace: Unified Text Operations

One of DocFrame' key innovations is the **text namespace** - a unified API that works across all Polars data types (Series, DataFrames, Expressions). The same `.text` methods work everywhere!

In [ ]:
# Text namespace works on Series, DataFrames, and Expressions uniformly!

print("🔤 Text Namespace Demonstration")
print("=" * 50)

# 1. On individual Series (traditional approach)
sample_text = tweets_df.document.head(1)
print("1. Text operations on Series:")
print(f"   Text: {sample_text.to_list()[0][:60]}...")
print(f"   Word count: {sample_text.text.word_count().to_list()[0]}")
print(f"   Character count: {sample_text.text.char_count().to_list()[0]}")
print()

# 2. On DataFrame expressions (DocFrame magic!)
print("2. Text operations on DataFrame (vectorized):")
text_stats = tweets_df.head(5).select(
    [
        pl.col("username"),
        pl.col("text"),
        pl.col("text").text.word_count().alias("words"),
        pl.col("text").text.char_count().alias("chars"),
        pl.col("text").text.sentence_count().alias("sentences"),
    ]
)
print(text_stats)
print()

# 3. Using the document shortcut (DocDataFrame convenience)
print("3. Using DocDataFrame's document property:")
quick_stats = tweets_df.head(5).select(
    [
        pl.col("username"),
        tweets_df.document.text.word_count().alias("words"),
        tweets_df.document.text.char_count().alias("chars"),
    ]
)
print(quick_stats)

## 4. Automatic Text Statistics

DocFrame provides built-in methods for comprehensive text analysis. Let's analyze our tweet dataset!

In [ ]:
# Get comprehensive text statistics using DocFrame' built-in method
print("📊 Comprehensive Text Statistics")
print("=" * 40)

# Use DocDataFrame's describe_text method for automatic statistics
text_stats = tweets_df.describe_text()
print("Dataset-wide text statistics:")
print(text_stats)
print()

# Add text statistics as new columns for further analysis
tweets_with_stats = tweets_df.with_columns(
    [
        pl.col("text").text.word_count().alias("word_count"),
        pl.col("text").text.char_count().alias("char_count"),
        pl.col("text").text.sentence_count().alias("sentence_count"),
    ]
)

print("Sample tweets with computed statistics:")
print(
    tweets_with_stats.head(3).select(
        ["username", "word_count", "char_count", "sentence_count", "text"]
    )
)
print()

# Analyze by user - who writes the longest tweets?
user_stats = (
    tweets_with_stats.group_by("username")
    .agg(
        [
            pl.count().alias("tweet_count"),
            pl.col("word_count").mean().alias("avg_words"),
            pl.col("char_count").mean().alias("avg_chars"),
            pl.col("word_count").max().alias("max_words"),
        ]
    )
    .sort("avg_words", descending=True)
)

print("Top 10 users by average words per tweet:")
print(user_stats.head(10))

## 5. Text Processing and Cleaning

DocFrame provides powerful text cleaning and processing capabilities. Let's clean our tweet data and extract insights!

In [ ]:
# Text cleaning with DocFrame
print("🧹 Text Cleaning and Processing")
print("=" * 35)

# Take a sample tweet and show cleaning steps
sample_tweet_df = tweets_df.head(1)
original_text = sample_tweet_df.document.to_list()[0]

print(f"Original tweet: {original_text}")
print()

# Apply various text cleaning operations using the text namespace
cleaned_df = sample_tweet_df.with_columns(
    [
        # Clean the text: normalize whitespace, remove extra spaces
        pl.col("text").text.clean().alias("cleaned"),
        # Tokenize into words
        pl.col("text").text.tokenize().alias("tokens"),
        # Extract word counts after cleaning
        pl.col("text").text.clean().text.word_count().alias("clean_word_count"),
    ]
)

print("Cleaning results:")
print(f"Cleaned text: {cleaned_df.get_column('cleaned').to_list()[0]}")
print(
    f"Tokens: {cleaned_df.get_column('tokens').to_list()[0][:10]}..."
)  # First 10 tokens
print(
    f"Word count after cleaning: {cleaned_df.get_column('clean_word_count').to_list()[0]}"
)
print()

# Apply cleaning to larger dataset for analysis
print("Processing larger dataset...")
cleaned_tweets = tweets_df.head(100).with_columns(
    [
        pl.col("text").text.clean().alias("cleaned_text"),
        pl.col("text").text.word_count().alias("original_words"),
        pl.col("text").text.clean().text.word_count().alias("cleaned_words"),
    ]
)

print("Cleaning effectiveness:")
cleaning_stats = cleaned_tweets.select(
    [
        pl.col("original_words").mean().alias("avg_original_words"),
        pl.col("cleaned_words").mean().alias("avg_cleaned_words"),
        (pl.col("original_words") - pl.col("cleaned_words"))
        .mean()
        .alias("avg_words_removed"),
    ]
)
print(cleaning_stats)

## 6. Document Column Management

DocFrame automatically detects your main text column, but you can easily switch between different text columns or manage multiple text fields.

In [ ]:
# Document column management features
print("📝 Document Column Management")
print("=" * 30)

# Show current document column detection
print(f"Current document column: '{tweets_df.active_document_name}'")
print(f"Available columns: {tweets_df.columns}")
print()

# Let's create a dataset with multiple text columns to demonstrate switching
multi_text_df = tweets_df.head(5).with_columns(
    [
        pl.col("text").alias("tweet_content"),
        pl.col("username").alias("user_handle"),
        # Create a summary column
        (
            pl.lit("Tweet by @")
            + pl.col("username")
            + pl.lit(": ")
            + pl.col("text").str.slice(0, 50)
            + pl.lit("...")
        ).alias("summary"),
    ]
)

print("Created dataset with multiple text columns:")
print(multi_text_df.select(["user_handle", "summary", "tweet_content"]).head(2))
print()

# Switch document column to different text field
doc_with_summary = DocDataFrame(multi_text_df.to_pandas(), document_column="summary")

print(f"Switched document column to: '{doc_with_summary.active_document_name}'")
print("Document property now refers to summary:")
for i, doc in enumerate(doc_with_summary.document.head(2).to_list()):
    print(f"{i + 1}. {doc}")
print()

# Demonstrate changing document column dynamically
doc_with_content = doc_with_summary.set_document("tweet_content")
print(f"Changed document column to: '{doc_with_content.active_document_name}'")
print("Now document property refers to tweet content:")
print(f"Sample: {doc_with_content.document.head(1).to_list()[0][:80]}...")
print()

# Show automatic detection for different datasets
print("Automatic detection on candidate data:")
print(f"Detected: '{candidates_df.active_document_name}'")
print("This chose 'evidence' because it has the longest average text length")

## 7. Advanced Text Analysis: Document-Term Matrix

DocFrame includes built-in support for creating document-term matrices (DTM) for further analysis like topic modeling or text classification.

In [ ]:
# Document-Term Matrix creation with DocFrame
print("🔢 Document-Term Matrix Analysis")
print("="*35)

# Take a small sample for DTM demonstration
sample_tweets = tweets_df.head(10).select(["username", "text"])
doc_sample = DocDataFrame(sample_tweets.to_pandas())

print(f"Creating DTM from {len(doc_sample)} tweets...")
print()

# Create DTM using DocFrame built-in method
dtm = doc_sample.to_dtm()

print(f"DTM shape: {dtm.shape}")
print(f"Vocabulary size: {len(dtm.columns)} unique terms")
print()

print("Sample of the DTM (first 5 rows, first 10 terms):")
print(dtm.head(5).select(dtm.columns[:10]))
print()

# Find most common terms across all documents
term_frequencies = dtm.sum().transpose(include_header=True)
term_freq_df = pl.DataFrame({
    "term": dtm.columns,
    "frequency": [dtm[col].sum() for col in dtm.columns]
}).sort("frequency", descending=True)

print("Top 15 most frequent terms across all tweets:")
print(term_freq_df.head(15))
print()

# Demonstrate text filtering and analysis workflow
print("Text filtering example:")
long_tweets = tweets_df.filter(
    pl.col("text").text.word_count() > 20
).head(5)

print(f"Found {len(long_tweets)} tweets with >20 words:")
for i, (user, text) in enumerate(long_tweets.select(["username", "text"]).iter_rows()):
    word_count = DocDataFrame(pl.DataFrame({"text": [text]})).document.text.word_count().to_list()[0]
    print(f"{i+1}. @{user} ({word_count} words): {text[:60]}..."))

## 8. Lazy Evaluation and Performance

DocFrame supports lazy evaluation through DocLazyFrame, giving you the performance benefits of Polars' lazy evaluation while maintaining text analysis capabilities.

In [ ]:
# Lazy evaluation with DocLazyFrame
print("⚡ Lazy Evaluation Performance")
print("=" * 32)

# Convert to lazy frame for efficient processing
tweets_lazy = DocLazyFrame(tweets_df.to_lazyframe())

print(
    f"Created DocLazyFrame with document column: '{tweets_lazy.active_document_name}'"
)
print("Lazy operations are built up but not executed until collect()")
print()

# Build a complex text analysis pipeline
analysis_pipeline = (
    tweets_lazy.with_columns(
        [
            # Add text statistics
            pl.col("text").text.word_count().alias("word_count"),
            pl.col("text").text.char_count().alias("char_count"),
            pl.col("text").text.sentence_count().alias("sentence_count"),
            # Clean and process text
            pl.col("text").text.clean().alias("clean_text"),
        ]
    )
    .filter(
        # Only tweets with substantial content
        (pl.col("word_count") >= 5) & (pl.col("char_count") >= 20)
    )
    .group_by("username")
    .agg(
        [
            pl.count().alias("tweet_count"),
            pl.col("word_count").mean().alias("avg_words"),
            pl.col("char_count").mean().alias("avg_chars"),
            pl.col("word_count").sum().alias("total_words"),
        ]
    )
    .filter(pl.col("tweet_count") >= 3)  # Users with at least 3 tweets
    .sort("avg_words", descending=True)
)

print("Built lazy analysis pipeline (not executed yet)")
print("Pipeline includes: text stats, cleaning, filtering, grouping, and sorting")
print()

# Execute the pipeline
print("Executing pipeline...")
results = analysis_pipeline.collect()

print(f"Results shape: {results.shape}")
print("Top 10 most verbose users (by average words per tweet):")
print(results.head(10))
print()

# Demonstrate lazy serialization
print("Lazy serialization:")
lazy_json = tweets_lazy.serialize()
print(f"Serialized lazy frame (first 100 chars): {lazy_json[:100]}...")

# Convert back to eager for final operations
final_df = DocDataFrame(results.to_pandas())
print(f"Final DocDataFrame created with {len(final_df)} users")

## 9. Data Integration and Serialization

DocFrame seamlessly integrates with the broader Python ecosystem and preserves all metadata during serialization.

In [ ]:
# Data integration and serialization
print("🔄 Data Integration and Serialization")
print("=" * 37)

# 1. Join datasets using document column information
print("1. Joining candidate and tweet data:")

# Create a small lookup of usernames to candidate info
candidate_lookup = candidates_df.select(
    ["username", "party", "gender", "first_name", "last_name"]
).filter(pl.col("username").is_not_null())

print(f"Candidate lookup: {len(candidate_lookup)} candidates with Twitter accounts")
print(candidate_lookup.head(3))
print()

# Join with tweet data to analyze by gender/party
enriched_tweets = tweets_df.head(1000).join(
    candidate_lookup, on="username", how="inner"
)

print(f"Enriched dataset: {len(enriched_tweets)} tweets from candidates")
print()

# Analyze tweet patterns by gender
gender_analysis = (
    enriched_tweets.group_by("gender")
    .agg(
        [
            pl.count().alias("tweet_count"),
            pl.col("text").text.word_count().mean().alias("avg_words_per_tweet"),
            pl.col("text").text.char_count().mean().alias("avg_chars_per_tweet"),
        ]
    )
    .sort("tweet_count", descending=True)
)

print("Tweet patterns by gender:")
print(gender_analysis)
print()

# 2. Serialization with metadata preservation
print("2. Serialization with metadata preservation:")

# Create a rich DocDataFrame with metadata
sample_for_serialization = enriched_tweets.head(5)
doc_sample = DocDataFrame(sample_for_serialization.to_pandas())

print(f"Original document column: '{doc_sample.active_document_name}'")

# Serialize to JSON (preserves document column info)
serialized = doc_sample.serialize(format="json")
print(f"Serialized size: {len(serialized)} characters")

# Deserialize and verify metadata is preserved
restored_df = DocDataFrame.deserialize(serialized, format="json")
print(f"Restored document column: '{restored_df.active_document_name}'")
print(f"Shape preserved: {restored_df.shape == doc_sample.shape}")
print(f"Columns preserved: {set(restored_df.columns) == set(doc_sample.columns)}")
print()

# 3. Export to standard formats while preserving analysis
print("3. Export capabilities:")
print("✓ Can export to CSV, Parquet, JSON while preserving computed text statistics")
print("✓ Document column information can be serialized and restored")
print("✓ Seamless integration with pandas, scikit-learn, and other tools")

# Example: prepare for machine learning
ml_ready = enriched_tweets.select(["text", "gender", "party"]).with_columns(
    [
        pl.col("text").text.word_count().alias("word_count"),
        pl.col("text").text.char_count().alias("char_count"),
        pl.col("text").text.sentence_count().alias("sentence_count"),
    ]
)

print(f"\\nML-ready dataset shape: {ml_ready.shape}")
print("Ready for classification, clustering, or other ML tasks!")

## 10. Summary: Why DocFrame?

DocFrame transforms text analysis from a complex, multi-step process into an intuitive, unified workflow.

### 🎯 Key Benefits Demonstrated:

1. **Automatic Intelligence**: No more guessing which column contains your text data
2. **Unified API**: Same `.text` methods work everywhere - Series, DataFrames, Expressions
3. **Performance**: Built on Polars for speed and memory efficiency
4. **Seamless Integration**: Works with your existing data science tools
5. **Rich Text Analytics**: Built-in statistics, cleaning, and processing capabilities

### 📊 What We Analyzed:
- **6,000+ political tweets** from Queensland election candidates
- **133 candidate profiles** with gender and party information
- **Text statistics** across different demographics
- **Document-term matrices** for advanced analysis
- **Lazy evaluation** for performance optimization

### 🚀 Next Steps:
DocFrame is perfect for:
- **Social media analysis** (Twitter, Reddit, forums)
- **Document processing** (news articles, papers, reports)
- **Survey analysis** (open-ended responses, feedback)
- **Content analysis** (reviews, comments, descriptions)
- **Research workflows** where text is a key data type

Try DocFrame in your next text analysis project and experience the difference!